proposals/p1964.md

```diff
@@ -0,0 +1,372 @@
+    modeled with normal string literals.
+
+## Details
+
+-   A character literal is a sequence enclosed with single quotes delimiter ('),
+    of UTF-8 code units that must be a valid encoding. This matches
+    [the UTF-8 encoding of Carbon source files](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0142.md#character-encoding).
+-   A character literal must encode exactly one code point.
+-   It supports addition and subtraction. These operations produce another code
+    point literal, or produce an error if the result is out of range.
+-   Character literals support some back-slash (`\`) escape sequences, including
+    `\t`, `\n`, `\r`, `\"`, `\'`, `\\`, `\0`, and `\u{HHHH...}`. See
+    [String Literals: Escape sequence](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0199.md#escape-sequences).
+
+We will not support:
+
+-   character literals that don't contain exactly one Unicode code point;
+-   multi-line literals;
+-   "raw" literals (using #'x'#);
+-   `\x` escape sequences;
+-   character literals with a single quote (`'`) or back-slash (`\`), except as
+    part of an escape sequence
+-   empty character literals (`''`);
+-   ASCII control codes (0...31), including whitespace characters other than
+    word space (tab, line feed, carriage return, form feed, and vertical tab),
+    except when specified with an escape sequence.
+
+### Types
+
+For the time being we will support a type `CharN` that will hold both code units
+and code points, and will leave the different UTF-encoding code unit types to
+another proposal. See
+[UTF code unit types proposal](#utf-code-unit-types-proposal))
+
+We will have the type `CharN` and only support literals that map directly to the
+complete value of a code point.
+
+```
+let allowed: CharN = 'a';
+```
+
+The above is allowed because the type of `'a'` is the character literal
+consisting of the single Unicode code point 97, which can be converted to
+`CharN` since 97 is less than or equal to 0x7F.
+
+### Operations
+
+Character literals representing a single code point support the following
+operators:
+
+-   Comparison: `<`, `>`, `<=`, `>=` `==`
+-   Plus: `+`. This doesn't concatenate, but allows numerically adjusting the
+    value:
+    -   Only one operand may be a character literal, the other must be an
+        integer literal.
+    -   The result is the character literal whose numeric value is the sum of
+        numeric value of the operands. If that sum is not a valid Unicode code
+        point, it is an error.
+-   Subtract: `-`. This will subtract the value of the two characters, or a
+    character followed by an integer literal:
+    -   If the `-` is used between two character literals, the result will be a
+        character constant `'z' - 4`.
+    -   If the `-` is used between a character literal, followed by a integer
+        literal this will produce an integer constant `'z' - 'a'`.
```

Comment:
```suggestion
    -   If the `-` is used between two character literals, the result will be an
        integer constant. For example, `'z' - 'a'` is equivalent to `25`.
    -   If the `-` is used between a character literal followed by a integer
        literal, this will produce a character constant. For example `'z' - 4`
        is equivalent to `'v'`.
```